llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp

```diff
@@ -1275,6 +1275,56 @@ void SelectionDAGLegalize::LegalizeOp(SDNode *Node) {
       }
     }
     break;
+    case ISD::FCANONICALIZE: {
+      const Triple &TT = DAG.getTarget().getTargetTriple();
+      if (TT.getArch() == Triple::x86 || TT.getArch() == Triple::x86_64) {
```

Comment:
Do not include architecture checks

Reply:
@arsenm Apart from the positive-zero handling, is there anything here that shouldn't be target-independent? Given how many targets already support canonicalize, I'm a bit surprised that there would be any new target-independent code needed. It seems that the other targets have pattern-matching lowering for the intrinsic. Is that preferable to the approach taken here?

Reply:
@andykaylor I was under the impression other targets except amdgpu also crashes, due to lack of implementation. To the best of my knowledge, I haven't come across any pattern that does load from constant pool, also I imagine vectors would be difficult to handle in such a case.

Reply:
The target specific cases would be x86_fp80 (and probably ppcf128), which are faux target independent

Reply:
Regarding the reason I guarded this with the arch check, is that, I originally Intended to do all this code in X86IselDagToDag, But when handling subnormal cases in selection phase I would run into all sorts of crashes and It became apparent that handling subnormal cases prior to selection is what I had to do.

 About the root cause of the crashes itself, when I would try to generate a load of zero from constant pool for a positive denormal for example ( in case of preserve-sign denormal mode ) it would generate mal-formed/illegal mir/dag node, take a look at the below example 

```
Correct /Expected
    t9: f32,ch = VMOVSSZrm_alt<Mem:(load (s32) from constant-pool)> Register:i64 $rip, TargetConstant:i8<1>, Register:i64 $noreg, TargetConstantPool:i32<float -1.000000e+00> 0, Register:i16 $noreg, t0
 
 Generated 
t19: f32,ch = VMOVSSZrm_alt<Mem:(load (s32) from constant-pool)> ConstantPool:i64<float -0.000000e+00> 0, TargetConstant:i8<1>, Register:i64 $noreg, TargetConstant:i32<0>, Register:i16 $noreg, t0

Root cause [ Register:i64 $rip ] not selected still the constant pool entry exists ConstantPool:i64<float -0.000000e+00> 0

```
As you can see the register operand rip, was not selected and the constant pool reference still existed post instruction selection, till code emission. Resulting in crash, and rightly so. 

About the guard itself, when I tested this same code without any arch guard I saw a test case in amdgpu was crashing named ```clamp.ll``` to prevent that I had to resort to adding this guard.

Reply:
Your patch is mechanically broken. You shouldn't need to do any of this manual replacement you are trying to do, or inspect the values of the incoming operand. Please remove all attempts at handling constants from this patch. Constant folding is an optimization that should be entirely separate from legalization and selection. None of this code should be specially considering constants.

A patch to support canonicalize should only handle converting the intrinsic to either a strict_ node in the legalizer (such as strict_fmul * 1.0, or strict_fadd -0), or directly selecting to a target instruction. For the constant case, you should just end up emitting a multiply of a constant by a constant (which can be optimized later, in a separate optimization combine)

Reply:
@arsenm You mean to say, doing subnormal canonicalization in DAG combine would be a better design. I agree, anywhere pre selection sounds fine, The reason I chose legalization phase is due to availability of ```ExpandConstantFP``` API, I haven't looked at combiner code, but If it's not available there, I might have to duplicate the API's functionality, unless there is no way to access this api from combine, should be doable, I reckon.

Reply:
If I understand correctly, you say, let's not handle constants/subnormals in this PR, and only handle variable canonicalization. And follow up with a PR that handles constants/subnormals in combiner. Am I right?

Reply:
You do not need ExpandConstantFP at all. You do not need to consider constants, regardless of the class. Legalization must work for all inputs, which covers constants materialized in registers

Reply:
Okay, Let's try to agree on the design first. 

- If x is a constant and denormal
  - If x is a denormal flush it to zero when denormal-fp-math=preserve-sign. And generate + or – zero according to the sign. 
  - If x is the denormal mode should never be PositiveZero for x86. That should be an error, Assert when denormal-fp-math=positive.
  - If x is a denormal with mode denormal-fp-math=ieee, Perform replacement like y = x.
  - If x is a denormal with mode denormal-fp-math=dynamic, Perform y = 1.0 * x. 
- If x is a constant and a signalling NaN, replace it with a quiet NaN. 
- If x is a regular fp constant like ( 3.14, 1.7, 2.30 … etc ) 
  - Simply do a copy y = x  or in other words keep it as it is. 
- If x is an undef replace with qnan.
- If x is a variable replace with y = x * 1.0

Reply:
You mean to say either 

Get rid of all this code from legalization and put it elsewhere say in say X86IselLowering/doing combines. I did that, look at the following code, and then select variables using tablegen 
```
SDValue combineConstantOrUndefCanonicalize(SDNode *Node,SelectionDAG &DAG){
  SDValue Operand = Node->getOperand(0);
  SDLoc dl(Node);
  EVT VT = Operand.getValueType();
  if (ConstantFPSDNode *CFP = dyn_cast<ConstantFPSDNode>(Operand)) {
    const APFloat &C = CFP->getValueAPF();
    if (C.isDenormal()) {
      DenormalMode Mode =
          DAG.getMachineFunction().getDenormalMode(C.getSemantics());
      assert((Mode != DenormalMode::getPositiveZero()) &&
              "Positive denormal mode is not valid for X86 target.");
      if (Mode == DenormalMode::getPreserveSign()) {
        SDValue SDZero =
            DAG.getConstantFP((C.isNegative() ? -0.0 : 0.0), dl, VT);        
        return SDZero;
      } else if (Mode == DenormalMode::getIEEE()) {
        return Operand;
      }
    } else if (C.isNaN() && C.isSignaling()) {
      APFloat CanonicalQNaN = APFloat::getQNaN(C.getSemantics());
      SDValue QuitNaN = DAG.getConstantFP(CanonicalQNaN, dl, VT);
      return QuitNaN;
    }
  }else if (Operand.isUndef()) {
    APFloat CanonicalQNaN =
        APFloat::getQNaN(VT.getFltSemantics());
    SDValue QuitNaN = DAG.getConstantFP(CanonicalQNaN, dl, VT);
    return QuitNaN;
  }
  return Operand;
}

```

**OR** 

use ISD::STRICT_FMUL/ISD::STRICT_FADD instead of ISD::FADD ( well this crashes saying unable to select strict_fmul ) but assuming that crash is resolved, but there is still then the above function ```combineConstantOrUndefCanonicalize``` in x86isellowering will not be reachable as selection will replace everything with mul. So then are you saying let it replace with mul ( whatever the case subnormals/consts/vars ) and later it will be optimized away or don't care whatever happens as long as mul remains. If not then continuing the train of thought from ```combineConstantOrUndefCanonicalize``` not being reached as selection and x86isellowering is tightly coupled, and selection will do it's thing as per this patch, in this case we will be unable to handle subnormals as intrinsic will be gone by then. 

**OR** 

Just do the following here itself in legalizer. 
```
    if (ConstantFPSDNode *CFP = dyn_cast<ConstantFPSDNode>(Operand)) {
          const APFloat &C = CFP->getValueAPF();
          if (C.isDenormal()) {
            DenormalMode Mode =
                DAG.getMachineFunction().getDenormalMode(C.getSemantics());
            assert((Mode != DenormalMode::getPositiveZero()) &&
                   "Positive denormal mode is not valid for X86 target.");
            if (Mode == DenormalMode::getPreserveSign()) {
              SDValue SDZero =
                  DAG.getConstantFP((C.isNegative() ? -0.0 : 0.0), dl, VT);
              NewNode = SDZero.getNode();
            } else if (Mode == DenormalMode::getIEEE()) {
              NewNode = Operand.getNode();
            }
          } else if (C.isNaN() && C.isSignaling()) {
            APFloat CanonicalQNaN = APFloat::getQNaN(C.getSemantics());
            SDValue QuitNaN = DAG.getConstantFP(CanonicalQNaN, dl, VT);
            NewNode = QuitNaN.getNode();
          }
        } else if (Operand.isUndef()) {
          APFloat CanonicalQNaN = APFloat::getQNaN(VT.getFltSemantics());
          SDValue QuitNaN = DAG.getConstantFP(CanonicalQNaN, dl, VT);
          NewNode = QuitNaN.getNode();
        }
        break;
      }
```

Reply:
Either 1) or 3) looks good to me, though I'd choose 1) if I do it.

2) is a general target independent method. The crash is probably due to you didn't handle input chain. As @andykaylor pointed, many targets have already supported canonicalize. There's no much value to do it in this way.

Reply:
Hi Phoebe, I agree, 1 looks like a cleaner implementation, The disadvantage there is every target will have duplicated code, Note that only AMDGPU has implemented the intrinsic, no other target. Refer https://github.com/llvm/llvm-project/pull/106370#discussion_r1735775923 

2 doesn't seem a clean implementation. Which is why I'm leaning towards 3, to avoid code duplication.

Reply:
If you search `fcanonicalize` in td files, you will find LoongArch/Mips/PowerPC also define it by patterns. Besides, AArach64 has a test for `llvm.canonicalize.f32`. So it looks like most mainstream targets support it to some extent.

This isn't really duplication. E.g., LoongArch/Mips use fmin/fmax to lower it. Maybe fmul/fadd is more expensive to them, but X86 cannot use fmin/fmax because it doesn't support signed zero.

Reply:
Yes In that case 1 is the way to go. Thanks for the reply.